api/turing/api/appcontext.go

```diff
@@ -81,19 +85,50 @@ func NewAppContext(
 		return nil, errors.Wrapf(err, "Failed initializing cluster controllers")
 	}
 
+	// Initialise Batch components
+	// Since there is only the default environment, we will not create multiple batch runners.
+	ensemblingImageBuilder, err := imagebuilder.NewEnsemberJobImageBuilder(
+		clusterControllers[cfg.EnsemblingJobConfig.DefaultEnvironment],
+		cfg.EnsemblingJobConfig.ImageConfig,
+		cfg.EnsemblingJobConfig.KanikoConfig,
+	)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed initializing ensembling image builder")
+	}
+
+	ensemblingJobService := service.NewEnsemblingJobService(db, cfg.EnsemblingJobConfig.DefaultEnvironment)
+	batchEnsemblingController := batchcontroller.NewBatchEnsemblingController(
+		clusterControllers[cfg.EnsemblingJobConfig.DefaultEnvironment],
+		mlpSvc,
+		cfg.SparkInfraConfig,
+	)
+
+	batchEnsemblingJobRunner := batchrunner.NewBatchEnsemblingJobRunner(
+		batchEnsemblingController,
+		ensemblingJobService,
+		mlpSvc,
+		ensemblingImageBuilder,
+		cfg.EnsemblingJobConfig.InjectGojekConfig,
+		cfg.EnsemblingJobConfig.DefaultEnvironment,
+		cfg.EnsemblingJobConfig.BatchSize,
+		cfg.EnsemblingJobConfig.MaxRetryCount,
+		cfg.EnsemblingJobConfig.ImageConfig.BuildTimeoutDuration,
+	)
+
 	appContext := &AppContext{
 		Authorizer:            authorizer,
 		DeploymentService:     service.NewDeploymentService(cfg, clusterControllers),
 		RoutersService:        service.NewRoutersService(db),
 		EnsemblersService:     service.NewEnsemblersService(db),
-		EnsemblingJobService:  service.NewEnsemblingJobService(db, cfg.EnsemblingJobConfig.DefaultEnvironment),
+		EnsemblingJobService:  ensemblingJobService,
 		RouterVersionsService: service.NewRouterVersionsService(db),
 		EventService:          service.NewEventService(db),
 		RouterDefaults:        cfg.RouterDefaults,
 		CryptoService:         cryptoService,
 		MLPService:            mlpSvc,
 		ExperimentsService:    expSvc,
 		PodLogService:         service.NewPodLogService(clusterControllers),
+		BatchRunners:          []batchrunner.BatchJobRunner{batchEnsemblingJobRunner},
```

Comment:
Ok, I think that makes sense.

Reply:
It's a little messy if I were to do it outside in the main function because now I need to export cluster controllers.

Reply:
The logic for the initialization could be moved to `api/turing/batch/runner/runner.go` and the function can be called from main.go, with the appropriate parameters. That should be cleaner, right?

Reply:
Not quite since there is a dependency on clusterControllers (https://github.com/gojek/turing/pull/61/files#diff-062afe4cc3963430883838990efd8db7bed1a9dff8dbdb16fb9374fbc8b4afc0R91), which kinda leads to the problem of either instantiating them again, or exporting the clusterControllers out, which doesn't seem to solve the problem we are trying to solve.

The way I see it with AppContext is that it is more of a DI kind of solution where we instantiate all our dependencies here.